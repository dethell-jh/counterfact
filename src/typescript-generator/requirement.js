export class Requirement {
  constructor(data, url = "", specification = undefined) {
    if (data === undefined) {
      throw new Error("cannot create requirement without data");
    }

    this.data = data;
    this.url = url;
    this.specification = specification;
  }

  get isReference() {
    return this.data?.$ref !== undefined;
  }

  reference() {
    return this.specification.requirementAt(this.data.$ref, this.url);
  }

  select(path, data = this.data, basePath = "") {
    const [head, ...tail] = path.split("/");

    const branch = data[this.unescapeJsonPointer(head)];

    if (!branch) {
      return undefined;
    }

    if (tail.length === 0) {
      return new Requirement(
        branch,
        `${this.url}/${basePath}${head}`,
        this.specification
      );
    }

    return this.select(tail.join("/"), branch, `${basePath}${head}/`);
  }

  toJSON() {
    return {
      specification: this.specification,
      url: this.url,
      data: this.data,
    };
  }

  forEach(callback) {
    Object.keys(this.data).forEach((key) => {
      callback([key, this.select(this.escapeJsonPointer(key))]);
    });
  }

  map(callback) {
    const result = [];

    // eslint-disable-next-line array-callback-return
    this.forEach((entry) => result.push(callback(entry)));

    return result;
  }

  escapeJsonPointer(string) {
    return string.replaceAll("~", "~0").replaceAll("/", "~1");
  }

  unescapeJsonPointer(pointer) {
    return pointer.replaceAll("~1", "/").replaceAll("~0", "~");
  }
}
