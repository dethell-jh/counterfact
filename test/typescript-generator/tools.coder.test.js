/* eslint-disable jest/no-restricted-matchers */
import { ToolsCoder } from "../../src/typescript-generator/tools-coder.js";
import { Requirement } from "../../src/typescript-generator/requirement.js";

describe("a ToolsCoder", () => {
  it("has all hard-coded return values", async () => {
    const coder = new ToolsCoder(new Requirement({}));

    const [one, two, three] = coder.names();

    expect([one, two, three]).toStrictEqual(["Tools", "Tools2", "Tools3"]);

    expect(coder.id).toBe("tools");
    expect(coder.modulePath()).toBe("internal/tools.ts");
    await expect(coder.delegate()).resolves.toBe(coder);

    expect(
      coder.write({
        importExternal(name, path) {
          coder.imported = { name, path };
        },
      })
    ).toMatchSnapshot();

    expect(coder.imported).toStrictEqual({
      name: "JSONSchema6",
      path: "json-schema",
    });
  });
});
