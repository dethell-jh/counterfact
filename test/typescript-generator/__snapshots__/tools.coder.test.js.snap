// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a ToolsCoder has all hard-coded return values 1`] = `
"
    {
        oneOf: (array: string[]) => string;
        randomFromSchema: (schema: JSONSchema6 ) => unknown;
        accepts: (contentType: string) => boolean;
    }"
`;
